flang/test/Lower/array-elemental-calls-char.f90

```diff
@@ -0,0 +1,220 @@
+! CHECK:   %[[VAL_81:.*]] = fir.convert %[[VAL_80]] : (i32) -> i8
+! CHECK:   %[[VAL_82:.*]] = fir.undefined !fir.char<1>
+! CHECK:   %[[VAL_83:.*]] = fir.insert_value %[[VAL_82]], %[[VAL_81]], [0 : index] : (!fir.char<1>, i8) -> !fir.char<1>
+! CHECK:   fir.store %[[VAL_83]] to %[[VAL_72]] : !fir.ref<!fir.char<1>>
+! CHECK:   %[[VAL_84:.*]] = fir.convert %[[VAL_72]] : (!fir.ref<!fir.char<1>>) -> !fir.ref<!fir.char<1,?>>
+! CHECK:   %[[VAL_85:.*]] = fir.emboxchar %[[VAL_84]], %[[VAL_71]] : (!fir.ref<!fir.char<1,?>>, index) -> !fir.boxchar<1>
+! CHECK:   %[[VAL_86:.*]] = fir.call @_QPelem(%[[VAL_85]]) : (!fir.boxchar<1>) -> i32
+! CHECK:   %[[VAL_87:.*]] = fir.array_coor %[[VAL_88]](%[[VAL_73]]) %[[VAL_77]] : (!fir.ref<!fir.array<10xi32>>, !fir.shape<1>, index) -> !fir.ref<i32>
+! CHECK:   fir.store %[[VAL_86]] to %[[VAL_87]] : !fir.ref<i32>
+! CHECK:   %[[VAL_89:.*]] = subi %[[VAL_75]], %[[VAL_71]] : index
+! CHECK:   br ^bb1(%[[VAL_77]], %[[VAL_89]] : index, index)
+! CHECK: ^bb3:
+! CHECK:   return
+  i = elem(char(j))
+end subroutine
+
+! CHECK-LABEL: func @_QMchar_elemPfoo4(
+! CHECK-SAME: %[[VAL_106:[^:]+]]: !fir.ref<!fir.array<10xi32>>,
+! CHECK-SAME: %[[VAL_103:[^:]+]]: !fir.ref<!fir.array<10xi32>>)
+subroutine foo4(i, j)
+  integer :: i(10), j(10)
+! CHECK-DAG:   %[[VAL_90:.*]] = constant 5 : index
+! CHECK-DAG:   %[[VAL_91:.*]] = constant 10 : index
+! CHECK-DAG:   %[[VAL_92:.*]] = constant 0 : index
+! CHECK-DAG:   %[[VAL_93:.*]] = constant 1 : index
+! CHECK:   %[[VAL_94:.*]] = fir.shape %[[VAL_91]] : (index) -> !fir.shape<1>
+! CHECK:   %[[VAL_95:.*]] = fir.address_of(@{{.*}}) : !fir.ref<!fir.char<1,5>>
+! CHECK:   br ^bb1(%[[VAL_92]], %[[VAL_91]] : index, index)
+! CHECK: ^bb1(%[[VAL_96:.*]]: index, %[[VAL_97:.*]]: index):
+! CHECK:   %[[VAL_98:.*]] = cmpi sgt, %[[VAL_97]], %[[VAL_92]] : index
+! CHECK:   cond_br %[[VAL_98]], ^bb2, ^bb3
+! CHECK: ^bb2:
+! CHECK:   %[[VAL_99:.*]] = fir.convert %[[VAL_95]] : (!fir.ref<!fir.char<1,5>>) -> !fir.ref<!fir.char<1,?>>
+! CHECK:   %[[VAL_100:.*]] = fir.emboxchar %[[VAL_99]], %[[VAL_90]] : (!fir.ref<!fir.char<1,?>>, index) -> !fir.boxchar<1>
+! CHECK:   %[[VAL_101:.*]] = addi %[[VAL_96]], %[[VAL_93]] : index
+! CHECK:   %[[VAL_102:.*]] = fir.array_coor %[[VAL_103]](%[[VAL_94]]) %[[VAL_101]] : (!fir.ref<!fir.array<10xi32>>, !fir.shape<1>, index) -> !fir.ref<i32>
+! CHECK:   %[[VAL_104:.*]] = fir.call @_QPelem2(%[[VAL_100]], %[[VAL_102]]) : (!fir.boxchar<1>, !fir.ref<i32>) -> i32
+! CHECK:   %[[VAL_105:.*]] = fir.array_coor %[[VAL_106]](%[[VAL_94]]) %[[VAL_101]] : (!fir.ref<!fir.array<10xi32>>, !fir.shape<1>, index) -> !fir.ref<i32>
+! CHECK:   fir.store %[[VAL_104]] to %[[VAL_105]] : !fir.ref<i32>
+! CHECK:   %[[VAL_107:.*]] = subi %[[VAL_97]], %[[VAL_93]] : index
+! CHECK:   br ^bb1(%[[VAL_101]], %[[VAL_107]] : index, index)
+! CHECK: ^bb3:
+! CHECK:   return
+  i = elem2("hello", j)
+end subroutine
+
+!! TODO: character return for elemental functions.
```

Comment:
Noted.

Reply:
This one requires a bit more of standard reading. It is not clear to me if it is legal to do something like:

```
character(n) function foo(n)
! ...
c = foo([1, 2, 3, 4])
```
Where an elemental character function result length depends on the parameters, which could different for every element.